### PR TITLE
Initial install of the go lang version the support-scheduler

### PR DESCRIPTION
**Pull request by xmlviking**
Initial PR to load in the support-scheduler into EdgeX project.
Features:  Support Scheduler proper, CUD for Create, Update and Delete support Schedules and ScheduleEvents.   Default scheduler "scrub operations" .   This project is a logical port of the existing java based support scheduler. 

TODO:  Enhance the support scheduler configuration schema/process using the new V2 config seed work available in the current master version of EdgeX Go code base.

TODO: Enhance the callback and router capabilities in the support scheduler.

### CODE REVIEW

File: pkg/clients/scheduler/client.go [hunk 1]
```diff
@@ -55,7 +56,9 @@ func GetSchedulerClient() SchedulerClient {
 
 // Function to get a schedule from the remote scheduler server
 func (schedulerClient *schedulerRestClient) QuerySchedule(id string) (models.Schedule, error) {
-	client := &http.Client{}
+	client := &http.Client{
+		Timeout: time.Second * 10,
```
Comment (jduranf) [inline pkg/clients/scheduler/client.go]: Why this timeouts? As far as I know, every micro service defines its timeouts at startup:
```
http.TimeoutHandler(nil, time.Millisecond*time.Duration(command.Configuration.Service.Timeout), "Request timed out")
```
Why override this setting with another hard coded value?

Comment (xmlviking) [reply pkg/clients/scheduler/client.go]: We had a hard coded value in the original code base.  Change can be made to utilize some of our existing timeouts :) . Noted.

**Comment by tsconn23** [comment]
@xmlviking Recommend you go ahead and squash your commits as part of your next update to the feature branch.

File: pkg/clients/scheduler/client.go [hunk 2]
```diff
@@ -13,6 +13,7 @@ import (
 	"io"
 	"io/ioutil"
 	"net/http"
+	"time"
```
Comment (tsconn23) [inline pkg/clients/scheduler/client.go]: Be advised that the GetSchedulerClient() function below is outdated and should be replaced in order to facilitate the endpoint monitor pattern. See NewDeviceClient() below for an example:
https://github.com/edgexfoundry/edgex-go/blob/master/pkg/clients/metadata/device.go#L66

File: cmd/support-scheduler/main.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+	flag.BoolVar(&useConsul, "consul", false, "Indicates the service should use consul.")
+	flag.BoolVar(&useConsul, "c", false, "Indicates the service should use consul.")
+	flag.StringVar(&useProfile, "profile", "", "Specify a profile other than default.")
+	flag.StringVar(&useProfile, "p", "", "Specify a profile other than default.")
+	flag.Usage = usage.HelpCallback
+	flag.Parse()
+
+	//Read Configuration.
+	configuration := &scheduler.ConfigurationStruct{}
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/main.go]: Will do :) .

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+Host = '127.0.0.1'
+ConsulPort = 8500
+ServiceTimeout = 5000
+ConsulHost = 'edgex-core-consul'
+ConsulCheckAddress = 'http://localhost:48085/api/v1/ping'
+CheckInterval = '10s'
+ServiceAddress = '127.0.0.1'
+AppOpenMsg = 'This is the Support Scheduler Micro Service'
+EnableRemoteLogging = false
+LogFile = './logs/edgex-core-support-scheduler.log'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: cmd/support-scheduler/res/configuration.toml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+Host = '127.0.0.1'
+ConsulPort = 8500
+ServiceTimeout = 5000
+ConsulHost = 'localhost'
+ConsulCheckAddress = 'http://127.0.0.1:48085/api/v1/ping'
+CheckInterval = '10s'
+ServiceAddress = '127.0.0.1'
+AppOpenMsg = 'This is the Support Scheduler Micro Service'
+EnableRemoteLogging = false
+LogFile = './logs/edgex-core-support-scheduler.log'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration.toml]: done

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+LoggingRemoteUrl = 'http://localhost:48061/api/v1/logs'
+ServiceLabels = ''
+ServiceCallback = '/v1/callbacks'
+ServiceConnectRetries = 6
+ServiceConnectInterval = 10000
+ScheduleInterval = 500
+DefaultScheduleServiceProtocol = 'http'
+DefaultScheduleServicePort = 48080
+DefaultSchedulerServiceAddress = '127.0.0.1'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ConsulPort = 8500
+ServiceTimeout = 5000
+ConsulHost = 'edgex-core-consul'
+ConsulCheckAddress = 'http://localhost:48085/api/v1/ping'
+CheckInterval = '10s'
+ServiceAddress = '127.0.0.1'
+AppOpenMsg = 'This is the Support Scheduler Micro Service'
+EnableRemoteLogging = false
+LogFile = './logs/edgex-core-support-scheduler.log'
+LoggingRemoteUrl = 'http://localhost:48061/api/v1/logs'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+ConsulProfilesActive = 'docker'
+ReadLimit = 100
+ServicePort = 48085
+Host = '127.0.0.1'
+ConsulPort = 8500
+ServiceTimeout = 5000
+ConsulHost = 'edgex-core-consul'
+ConsulCheckAddress = 'http://localhost:48085/api/v1/ping'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+CheckInterval = '10s'
+ServiceAddress = '127.0.0.1'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: cmd/support-scheduler/res/configuration-docker.toml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+DefaultScheduleFrequency = 'P1D'
+DefaultScheduleStart = '20180101T000000'
+DefaultScheduleEventName = 'scrub-pushed-events,scrub-aged-events'
+DefaultScheduleEventMethod = 'DELETE,DELETE'
+DefaultScheduleEventService = 'core-data,core-data'
+DefaultScheduleEventParameters = ','
+DefaultScheduleEventPath = '/api/v1/event/scrub,/api/v1/event/removeold/age/604800000'
+DefaultScheduleEventSchedule = 'midnight,midnight'
+DefaultScheduleEventScheduler = 'support-scheduler,support-scheduler'
+MetaAddressableURL = 'http://localhost:48081/api/v1/addressable'
```
[the remark this answers is not in the document]
Comment (xmlviking) [reply cmd/support-scheduler/res/configuration-docker.toml]: done

File: pkg/clients/scheduler/client.go (hunk 1 above)
Comment (xmlviking) [reply pkg/clients/scheduler/client.go]: updated to use the configuration supplied ServiceTimeout.

**Comment by tsconn23** [comment]
recheck

**Comment by jduranf** [comment]
@xmlviking first of all, very nice work here!

Some comments:

1. I miss not using configuration v2. This new configuration system makes more easy to understand the toml files and implement some features like predefined schedules.
2. I don't agree with the initialization system, when adding the schedules defined in the configuration file and then creating the ID. I think the steps to follow should be:
- Request to core-metadata the schedules associated with support-scheduler and add this schedules to  the internal cache
- Check if any of the schedules defined in the configuration file are in the internal cache. If not, add this schedule in core-metadata and the use the returned ID
3. You have the assumption that all days have 24 hours, all months 30 days and all years 365 days. Unfortunately this is not true. Implementing the right behaviour is not easy, but at least should be fine to have a TODO.

**Comment by xmlviking** [comment]
Closing this PR and reopening on a new fork

**Pull request closed without merging**